File: course_101/exercises/easy_6/cute_angles.rb
# Write a method that takes a floating point number that represents an angle between 0 and 360 degrees and returns a String that represents that angle in degrees, minutes and seconds.
# You should use a degree symbol (°) to represent degrees, a single quote (') to represent minutes, and a double quote (") to represent seconds.
# A degree has 60 minutes, while a minute has 60 seconds.

DEGREE = "\xC2\xB0"
ARC_MINUTE = 60
ARC_SECOND = 60

def dms(degrees_float)
  total_seconds = degrees_float * ARC_MINUTE * ARC_SECOND
  degree, rest_seconds = total_seconds.divmod(ARC_MINUTE * ARC_SECOND)
  arc_minutes, arc_seconds = rest_seconds.divmod(ARC_MINUTE)

  "#{degree}#{DEGREE}#{format("%02d'%02d\"", arc_minutes, arc_seconds)}"
end

puts dms(30) == %(30°00'00")
puts dms(76.73) == %(76°43'48")
puts dms(254.6) == %(254°36'00")
puts dms(0) == %(0°00'00")
puts dms(93.034773) == %(93°02'05")
puts dms(360) == %(360°00'00") || dms(360) == %(0°00'00")

File: course_js210_computational_thinking_and_javascript_programming/exercises/js210_small_problems/easy_4/cute_angles.js
// Write a function that takes a floating point number representing an angle between 0 and 360 degrees, and returns a string representing that angle in degrees, minutes, and seconds. You should use a degree symbol (˚) to represent degrees, a single quote (') to represent minutes, and a double quote (") to represent seconds. There are 60 minutes in a degree, and 60 seconds in a minute.
// Note: your results may differ slightly depending on how you round values, but should generally be within a second or two of the results shown.
/*
PEDAC -------
Problem:
  Input:
    - Number
      - represents degrees

  Output:
    - String
      - degrees(°), minutes(') and seconds(")
  Rules:
    - 1° = 60minutes
    - 1 minute = 60 seconds

Data Structure:

Algorithm:
    Example input: E
  - get degrees
    - whole angle = degrees
  - get minutes
    - 60 * fraction rounded down
  - get seconds
    - fraction of minutes * 60 rounded down
  - return degrees + minutes + seconds
*/

const MINUTES_PER_DEGREE = 60;
const SECONDS_PER_MINUTE = 60;

function dms(angle) {
  let degrees = angle;
  let minutes = MINUTES_PER_DEGREE * (angle % 1);
  let seconds = SECONDS_PER_MINUTE * (minutes % 1);
  return `${Math.floor(degrees)}°${Math.floor(minutes).toPrecision(2)}'${Math.floor(seconds).toPrecision(2)}"`
}


console.log(dms(30));           // 30°00'00"
console.log(dms(76.73));        // 76°43'48"
console.log(dms(254.6));        // 254°35'59"
console.log(dms(93.034773));    // 93°02'05"
console.log(dms(0));            // 0°00'00"
console.log(dms(360));          // 360°00'00" or 0°00'00"
